Bug 1187979 - Improve Inputpad's touch and visual feeback

cr opened the pull request:
Inputpad looks and feels wonky compared to the slick look and feel of
the keyboard panel used while setting the pass code in Settings. Much of
the reason comes from that it only vibrates on key release, the other
that it does not behave well when the user is sliding a finger over the
keys. There is also a considerable delay between touching a key and a
visual reaction to the extent that it's skipped when typing mildly fast.

This prototype patch addresses all of the above:
- make InputPad vibrate instantly on touch
- make InputPad respect user's touchmove
- visualize key presses instantly

Commit:
Improve visual and touch feedback of Inputpad

Inputpad looks and feels wonky compared to the slick lock and feel of
the keyboard panel used while setting the pass code in Settings. Much of
the reason comes from that it only vibrates on key release, the other
that it does not behave well when the user is sliding a finger over the
keys. There is also a considerable delay between touching a key and a
visual reaction to the extent that it's skipped when typing mildly fast.

This patch addresses all of the above:

- make InputPad vibrate instantly on touch
- make InputPad respect user's touchmove
- visualize key presses instantly

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=24fd1bf25e3deb6e204e3fe2ce4d69daa8a4e5d6

Commit:
Fix DOM node comparison

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d03ea06eac1cb6cc4bda67d59c10025715ce200e

Commit:
Improving InputPad tests

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=074b856985795c6516479d6901bfd7c27d9570ff

cr commented:
Rebased to today's master, local tests passing.

Commit:
Linter fix for Inputpad test

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=04db3f4b74ae472f906fb500aa45f2fd5f3d22be

Commit:
Prevent vibrate on pending error timeout

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=226afb0d2a85f2751d4d8c7f269950277711f36a

Commit:
Work around eslint whining about unsafe innerHTML

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c885d13f60dd2cea471a47045ae226a25bf18fbc

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -111,25 +116,111 @@
+      case 'touchstart':
+        // TODO: ignore while on error timeout
+        touch = evt.changedTouches[0];
+        target = document.elementFromPoint(touch.clientX, touch.clientY);
```
Is it possible to use `evt.target` instead of querying element from the coordination?

https://developer.mozilla.org/en-US/docs/Web/Events/touchstart

cr replied:
I started out with .target, but unlike changedTouches it does not stay in sync with where the finger is touching on touch moves. IIRC, .target remains where the touch first went down. Hence I went for the alternative that always works.

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -111,25 +116,111 @@
+        }
+        break;
+      case 'touchmove':
+        // TODO: ignore while on error timeout
```
Is this still a TODO or just a comment need to be cleared?

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -111,25 +116,111 @@
+      case 'touchmove':
+        // TODO: ignore while on error timeout
+        touch = evt.changedTouches[0];
+        target = document.elementFromPoint(touch.clientX, touch.clientY);
```
`evt.target` ?

cr replied:
On touchmove, .target fails to report the current position under the finger. See comment above.

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -111,25 +116,111 @@
+        touch = evt.changedTouches[0];
+        target = document.elementFromPoint(touch.clientX, touch.clientY);
+        key = this._keyForTarget(target);
+        if (target.textContent !== this.lastTouched.textContent) {
```
Could you add some comments to make the intention here more clear?

snowmantw replied:
I mean, an overall comment to describe what you're doing here, with these conditional branches.

cr replied:
Added comments throughout the code.

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -47,8 +47,12 @@
     this.passcodeCode = document.getElementById('lockscreen-passcode-code');
     this.passcodePad = document.getElementById('lockscreen-passcode-pad');
     this.emergencyCallBtn = this.passcodePad.querySelector('a[data-key=e]');
+    this.lastTouched = null;
```
May need to describe what the variable is for in brevity.

cr replied:
Added this:
```
    /* lastTouched
     * Keeps track of the key that was under the finger during
     * the previous 'touchmove' event. UI must be redrawn only if
     * the finger moved to a different key.
     */
    this.lastTouched = null;
```

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -111,25 +116,111 @@
-           ) {
-          key = evt.target.parentNode.dataset.key;
+        // TODO: ignore while on error timeout
+        if (evt.type == 'touchend') {
```
No `==` in Gaia is a hinting requirement.

cr replied:
Code was deleted.

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -111,25 +116,111 @@
-          key = evt.target.parentNode.dataset.key;
+        // TODO: ignore while on error timeout
+        if (evt.type == 'touchend') {
+          evt.preventDefault();  // prevent the 'click'
```
Instead of calling `preventDefault` just for distinguishing these two events, to deal with these two events by two `switch` branch is better to me.

snowmantw replied:
And in such case if you concern the duplicated part, make them as a function and call it twice. Or, with such tiny duplicated code is okay I think.

cr replied:
I chose to go for the minor duplication, because else I feel that I should be writing tests for the new function. A future iteration may optimize the code in this respect.

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -111,25 +116,111 @@
+        if (evt.type == 'touchend') {
+          evt.preventDefault();  // prevent the 'click'
+          touch = evt.changedTouches[0];
+          target = document.elementFromPoint(touch.clientX, touch.clientY);
```
`evt.target` ?

cr replied:
Conscious design. See comment above.

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -111,25 +116,111 @@
+          target = evt.target;
+        }
+        key = this._keyForTarget(target);
+        if (this.lastTouched) {
```
I still feel it is unclear and need a comment to describe what's the relation between a `lastTouched` and key active/inactive. Maybe re-naming can solve that, but a comment may be better.

cr replied:
Added comments throughout the code. Refactored the variable to .lastTouchedKey.

In short: it's used to determine whether the finger has moved to a new key since the last event. If and only if that is the case, lastTouchedKey must be made inactive, and the new key must be made active.

Commit:
Added comments and gweng's changes

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b7752fc791d9b342c4c7ce4eeca8b4e9b942c31a

Commit:
Rewriting event handling and helper functions

This is a minor rewrite of InputPad's event handling and a major
rewrite of its helper functions. The event handler now delegates do
separate handlers for each touch/click event, creating a more compact
function design. Much of the internal state logic has been moved away
to be encapsulated in the helper functions.

The set of helper functions was restructured and rewritten to capture
more shared code between event handlers, to have consistent arguments,
and with unit testing in mind. Helpers now consistently take keypad
anchor elements as arguments where possible. They were also rewritten
to properly handle null arguments which result when events land outside
of the keypad.

Commit:
Adapting and improving InputPad tests

This is a major rewrite of InputPad tests, mostly triggered by the
restructuring. It also adds tests for previously uncovered code like
the keyi input handling, password checking, the emergency dialer, and
event handling in general.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=33e44b57a00a0c843ead79b5abd8a32b54600cbe

Commit:
Fixing linter whine

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=aca1184c937b994d959d1effbfa0fb3ec9884ec6
